### PR TITLE
Resolves mutant species keeping old organs they shouldn't have

### LINKED ISSUE #73449: Cardboard golems have tongues again
Reporting client version: 514.1589

## Round ID:
[200191](https://scrubby.melonmesa.com/round/200191)

## Testmerges:
- [Optimizes INVOKE_ASYNC by making it a macro and avoiding a proc call](https://www.github.com/tgstation/tgstation/pull/73264)

## Reproduction:
Talk as a cardboard golem/look at their organs

### LINKED ISSUE #73640: Plasmamen Have Hearts Again... Somehow
## Reproduction:

Yeah uh I don't think plasmamen are meant to have hearts at all. They have no blood, they haven't had hearts before and I can't see any PRs via searching that give plasmamen their hearts back. (Only one I can see is a 2020 pr that specifically REMOVED hearts from bloodless species.)

1. Open up a plasmaman via organ manipulation.
2. The plasmaman has a heart, which it shouldn't.

### LINKED ISSUE #74075: Abductors Spawn with Unintended Organs
Reporting client version: 514.1589

## Round ID:
[202010](https://scrubby.melonmesa.com/round/202010)

## Reproduction:
Found in 202006 - Abductors spawn with some organs from the players default char setup, including heart, lungs, and stomach (found because a former ethereal took abductor role and crystalized when they were killed).  Tested in round listed as well as locally to trigger/take abductor roles

Test 1 - Spawn in an abductor mob and check organs.  Spawned mobs do not have heart, lungs, or stomachs.  Spawn in an ethereal and then use VV to "Set Species" to abductor, then VV again.  Mob still contains ethereal organs (this is true of other mobs with unique organs in place of the 'lacking' abductor organs...plasmemes, flys, etc).

Test 2 (local?)- Set character to Ethereal (just because its easy to see the unique organs) in character setup, then observe the round.  Trigger abductors (needs 2 people, of course), and take the ghost role.  You'll get spawned in as an abductor, but you will have ethereal organs.  If you are killed, you'll crystalize and revive, just like an ethereal will, eventhough you've never been an ethereal in the round.  It seems to be pulling from the character you've selected on the char setup screen when spawning in this way.

As a note, Ethereals are likely the most impactful in this case, because of the Crystal core allowing for self revival.  Plasmeme lungs, for instance, will not impact the abductor as they have no_breath.  (At least as far as I can tell thus far).

Ethereal turned Abductor
image

Spawned Abductor using Spawn command
image

Event Triggered Abductor - Taken by observer with ethereal char selected in char setup
image

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

regenerate organs never removed invalid organs when regenerating
## Why It's Good For The Game

Fixes https://github.com/tgstation/tgstation/issues/73640
adds a unit test to prevent regressions in the future

## Changelog
:cl:
fix: Plasmamen don't have hearts, again
/:cl:

### CODE REVIEW

**Comment by ZephyrTFA** [comment]
> Can this get a test?

done

**Comment by ATH1909** [comment]
is this gonna cause weirdness for people who change species while they have, say, upgraded organs from robotics? a toolset implant? mutant organs from that one genetics machine?

File: code/modules/unit_tests/mutant_organs.dm [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		ORGAN_SLOT_BRAIN,
+		ORGAN_SLOT_HEART,
+		ORGAN_SLOT_LUNGS,
+		ORGAN_SLOT_EYES,
+		ORGAN_SLOT_EARS,
+		ORGAN_SLOT_TONGUE,
+		ORGAN_SLOT_LIVER,
+		ORGAN_SLOT_STOMACH,
+		ORGAN_SLOT_APPENDIX,
+		)
```
Comment (san7890) [inline code/modules/unit_tests/mutant_organs.dm:20]:
```suggestion
	)
```

File: code/modules/unit_tests/mutant_organs.dm [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+	for(var/datum/species/species_type as anything in species)
+		// change them to the species
+		dummy.set_species(species_type)
+
+		// check all their organs
+		for(var/organ_slot in organs_we_care_about)
+			var/expected_type = slot_to_species_organ_type(organ_slot, species_type)
+			var/obj/item/organ/actual_organ = dummy.getorganslot(organ_slot)
+			if(!actual_organ)
```
Comment (san7890) [inline code/modules/unit_tests/mutant_organs.dm:30]: could we use `isnull()` here (and elsewhere) instead?

**Review by san7890 — CHANGES_REQUESTED** [review]
also your CI is failing don't know if you noticed

File: code/modules/unit_tests/mutant_organs.dm [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
```
Comment (ZephyrTFA) [inline code/modules/unit_tests/mutant_organs.dm:30]:
```suggestion
			if(isnull(actual_organ))
```

File: code/modules/unit_tests/mutant_organs.dm [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+				if(expected_type)
+					TEST_FAIL("[species_type] did not update their [organ_slot] organ to [expected_type], no organ was found")
+					continue
+			else
+				if(!expected_type)
```
Comment (ZephyrTFA) [inline code/modules/unit_tests/mutant_organs.dm:35]:
```suggestion
				if(isnull(expected_type))
```

File: code/modules/unit_tests/mutant_organs.dm [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	for(var/datum/species/species_type as anything in species)
+		// change them to the species
+		dummy.set_species(species_type)
+
+		// check all their organs
+		for(var/organ_slot in organs_we_care_about)
+			var/expected_type = slot_to_species_organ_type(organ_slot, species_type)
+			var/obj/item/organ/actual_organ = dummy.getorganslot(organ_slot)
+			if(!actual_organ)
+				if(expected_type)
```
Comment (ZephyrTFA) [inline code/modules/unit_tests/mutant_organs.dm:31]:
```suggestion
				if(!isnull(expected_type))
```

**Comment by san7890** [comment]
image

```txt
 [00:49:41] Runtime in _external_organs.dm,357: bad color
    proc name: color image (/datum/bodypart_overlay/mutant/pod_hair/color_image)
    src: /datum/bodypart_overlay/mutant... (/datum/bodypart_overlay/mutant/pod_hair)
    call stack:
    /datum/bodypart_overlay/mutant... (/datum/bodypart_overlay/mutant/pod_hair): color image(, -6, the robotic head (/obj/item/bodypart/head/robot))
    /datum/bodypart_overlay/mutant... (/datum/bodypart_overlay/mutant/pod_hair): get overlay(-6, the robotic head (/obj/item/bodypart/head/robot))
    the robotic head (/obj/item/bodypart/head/robot): get limb icon(null, null)
    the robotic head (/obj/item/bodypart/head/robot): get limb icon(null, null)
    John Doe (/mob/living/carbon/human/consistent): update body parts(null)
    the podperson hair (/obj/item/organ/external/pod_hair): Insert(John Doe (/mob/living/carbon/human/consistent), 1, 0)
    Podperson (/datum/species/pod): regenerate organs(John Doe (/mob/living/carbon/human/consistent), Plasmaman (/datum/species/plasmaman), 1, null, 0)
    Podperson (/datum/species/pod): on species gain(John Doe (/mob/living/carbon/human/consistent), Plasmaman (/datum/species/plasmaman), 0)
    John Doe (/mob/living/carbon/human/consistent): set species(/datum/species/pod (/datum/species/pod), 1, 0)
    John Doe (/mob/living/carbon/human/consistent): set species(/datum/species/pod (/datum/species/pod), 1, 0)
    /datum/unit_test/mutant_organs (/datum/unit_test/mutant_organs): Run()
    RunUnitTest(/datum/unit_test/mutant_organs (/datum/unit_test/mutant_organs), /list (/list))
    RunUnitTests()
    /datum/callback (/datum/callback): InvokeAsync()
```

i wonder why this is fucking us over

**Comment by ZephyrTFA** [comment]
merging master and hoping that resolves it

File: code/modules/unit_tests/mutant_organs.dm [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/**
+ * Iterates over all species to ensure that organs are valid after being set to a mutant species
+ */
+/datum/unit_test/mutant_organs
+
+/datum/unit_test/mutant_organs/Run()
+	var/mob/living/carbon/human/consistent/dummy = allocate(/mob/living/carbon/human/consistent)
+	var/list/species = subtypesof(/datum/species)
+	species -= /datum/species/dullahan // the head is detached from the body
```
Comment (Kapu1178) [inline code/modules/unit_tests/mutant_organs.dm:9]:
```suggestion
	var/list/ignore = list(
		/datum/species/dullahan,
	)
	var/list/species = subtypesof(/datum/species) - ignore
```

**Pull request merged**